enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,359 @@
+
+If neither `.spec.tls.externalCertificate` or `.spec.tls.certificate` and `.spec.tls.key` are
+provided the router will serve the default generated certificates.
+
+All valid and invalid scenarios will be depicted via the existing `RouteIngressCondition`.
+
+#### Variation
+
+N.A
+
+### Implementation Details/Notes/Constraints [optional]
+
+The router will read the secret referenced in `.spec.tls.externalCertificate` if present and
+if the following pre-conditions (validated in the API server) are met it uses
+this certificate us configure haproxy.
+
+The router will bootstrap a watch based secret manager to ensure it can keep the
+certificate/secret up-to-date. This means the router pod will maintain active watches
+for every secret that is referenced by a route.
+
+Every active watch will be linked to a route, meaning the watch based secret manager
+will be linked to the lifecycle of the route. For every new route that is created,
+the secret manager will start a watch if the route uses `.spec.tls.externalCertificate`.
+For every update route event, the secret manager only increments the reference count.
+If a route is deleted, the secret manager will unregister the route and teardown the
+watch associated with it.
+
+The `ServiceAliasConfig` creation logic will be updated in the router to also parse
+the secret referenced in `.spec.tls.externalCertificate`. The router will
+use the default certificates only when `.spec.tls.certificate` or `.spec.tls.externalCertificate`
+are not provided.
+
+Validations done by the router as part of [ExtendedValidateRoute()](https://github.com/openshift/router/blob/c407ebbc5d8d85daea2ef2d1ba539444a06f4d25/pkg/router/routeapihelpers/validation.go#L158) (contents of secret),
+
+- Verify certificate and key (PEM encode/decode)
+- Verify private key matches public certificate
+
+Validations done by API server as part of [ValidateRoute()](https://github.com/openshift/openshift-apiserver/blob/aac3dd5bf0547e928103a0f718ca104b1bb13930/pkg/route/apis/route/validation/validation.go#L21),
+
+- The secret created should be in the same namespace as that of the route.
+- The secret created is of type `kubernetes.io/tls`.
+- The router serviceaccount must have permission to read this secret particular secret.
+  - The role and rolebinding to provide this access must be provided by the user.
+- CEL validations will enforce that both `.spec.tls.certificate` and `.spec.tls.externalCertificate`
+  are not specified on the route.
+
+### Risks and Mitigations
+
+There is a possibility of an invalid route being processed by the router (edge case),
+if any changes are done to the rbac or the referenced secret is deleted after the API
+server validation but before router has processed the route (maybe router pod is not running)
+then this can lead to the router processing this incorrect route.
+
+> Will need to duplicate the validations present on the API server to the router.
```

Comment:
Which doesn't eliminate the race, it just shortens the race window. Is it worth the effort? If the rbac or the secret is missing at the time the router needs them, I suppose the route's status will be updated. I think that's the best effort we can do in such a unfortunate scenario.

Reply:
I just realized that the route's status won't be updated with the missing RBAC condition because if the route is created than it was accepted by the api server. I agree with this statement then, we have to check the same in the router.

Reply:
We can resolve this conversation.